File: implementations/typescript/src/transport.ts
/**
 * Transport implementations — drop-in replacements for MCP SDK transports.
 *
 * Each transport implements the `Transport` interface from
 * `@modelcontextprotocol/sdk` and performs automatic LUMEN negotiation
 * on `start()`. If the remote peer doesn't respond to the LUMEN probe
 * within a configurable timeout, the transport falls back to JSON-RPC
 * transparently.
 */

import { spawn, ChildProcess } from "child_process";
import { createInterface, Interface } from "readline";
import { buildProbe, parseAck, DEFAULT_PROBE_TIMEOUT_MS } from "./negotiation.js";
import {
  buildFrame,
  buildSize,
  parseFrame,
  isCompressed,
  TYPE_REQUEST,
  TYPE_RESPONSE,
  TYPE_NOTIFY,
  FLAG_COMPRESSED,
} from "./frame.js";
import {
  FrameAssembler,
} from "./frame-assembler.js";
import { compressValue, decompressValue } from "./compress.js";

// ═══ Transport Interface (compatible with @modelcontextprotocol/sdk) ══════

/** Minimal transport interface matching MCP SDK's `Transport`. */
export interface Transport {
  start(): Promise<void>;
  send(message: JsonRpcMessage): Promise<void>;
  close(): Promise<void>;
  onmessage: ((message: JsonRpcMessage) => void) | null;
  onerror: ((error: Error) => void) | null;
  onclose: (() => void) | null;
}

/** JSON-RPC 2.0 message. */
export interface JsonRpcMessage {
  jsonrpc: "2.0";
  id?: string | number | null;
  method?: string;
  params?: Record<string, unknown>;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
}

// ═══ Options ════════════════════════════════════════════════════════════════

export interface LumenTransportOptions {
  /** Skip LUMEN negotiation entirely; use JSON-RPC directly. */
  forceJsonRpc?: boolean;
  /** Probe timeout in milliseconds (default: 500). */
  probeTimeoutMs?: number;
}

export interface LumenStdioOptions extends LumenTransportOptions {
  /** Command to spawn the MCP server process. */
  command: string;
  /** Arguments for the command. */
  args?: string[];
  /** Environment variables. */
  env?: Record<string, string>;
  /** Working directory. */
  cwd?: string;
}

// ═══ Stdio Transport ════════════════════════════════════════════════════════

/**
 * Drop-in replacement for `StdioClientTransport` from `@modelcontextprotocol/sdk`.
 *
 * On `start()`, spawns the MCP server process and attempts LUMEN negotiation.
 * If the server responds with a LUMEN ACK frame, all subsequent communication
 * uses LUMEN binary frames. Otherwise, falls back to JSON-RPC.
 *
 * @example
 * ```typescript
 * import { LumenStdioTransport } from "@lumen/mcp-transport";
 *
 * const transport = new LumenStdioTransport({
 *   command: "npx",
 *   args: ["-y", "@anthropic/mcp-server-filesystem", "/tmp"],
 * });
 * await transport.start();
 * ```
 */
export class LumenStdioTransport implements Transport {
  private proc: ChildProcess | null = null;
  private rl: Interface | null = null;
  private useLumen = false;
  private assembler = new FrameAssembler();

  onmessage: ((message: JsonRpcMessage) => void) | null = null;
  onerror: ((error: Error) => void) | null = null;
  onclose: (() => void) | null = null;

  constructor(private options: LumenStdioOptions) {}

  // ── Lifecycle ──────────────────────────────────────────────────────────

  async start(): Promise<void> {
    const opts = this.options;

    this.proc = spawn(opts.command, opts.args ?? [], {
      cwd: opts.cwd,
      env: { ...process.env, ...opts.env },
      stdio: ["pipe", "pipe", "pipe"],
    });

    this.proc.on("error", (err) => {
      this.onerror?.(err);
    });

    this.proc.on("exit", (code, signal) => {
      if (code !== 0) {
        this.onerror?.(
          new Error(`Server exited with code ${code}, signal ${signal}`),
        );
      }
      this.onclose?.();
    });

    // Stderr passthrough for debugging
    this.proc.stderr?.on("data", (chunk: Buffer) => {
      process.stderr.write(`[mcp-server stderr] ${chunk.toString()}`);
    });

    // Setup line reader for JSON-RPC path
    this.rl = createInterface({ input: this.proc.stdout! });

    // ── LUMEN negotiation ────────────────────────────────────────────
    if (!opts.forceJsonRpc) {
      const probe = buildProbe();
      this.proc.stdin!.write(probe);

      const negotiated = await this.waitForAck(
        opts.probeTimeoutMs ?? DEFAULT_PROBE_TIMEOUT_MS,
      );
      if (negotiated) {
        this.useLumen = true;
        // Switch to binary frame reader
        this.rl.close();
        this.startLumenReader();
        return;
      }
      // Fallback: server didn't ACK, but we already sent the probe bytes.
      // JSON-RPC servers will ignore them as garbage.
    }

    // ── JSON-RPC path ─────────────────────────────────────────────────
    this.rl.on("line", (line: string) => {
      try {
        const msg = JSON.parse(line) as JsonRpcMessage;
        this.onmessage?.(msg);
      } catch {
        // Ignore malformed lines (including binary probe garbage)
      }
    });
  }

  /** Wait for PROBE_ACK frame from stdout within timeout. */
  private waitForAck(timeoutMs: number): Promise<boolean> {
    return new Promise((resolve) => {
      const timer = setTimeout(() => resolve(false), timeoutMs);

      const onData = (chunk: Buffer) => {
        const frames = this.assembler.push(
          new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength),
        );

        for (const frame of frames) {
          if (frame.frameType === 0x10 /* TYPE_PROBE_ACK */) {
            this.proc!.stdout!.removeListener("data", onData);
            clearTimeout(timer);
            const ack = parseAck(frame.payload);
            resolve(ack !== null);
            return;
          }
        }
      };

      this.proc!.stdout!.on("data", onData);
    });
  }

  /** After LUMEN negotiation, read binary frames from stdout. */
  private startLumenReader(): void {
    this.proc!.stdout!.on("data", (chunk: Buffer) => {
      const frames = this.assembler.push(
        new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength),
      );

      for (const frame of frames) {
        let payload: Record<string, unknown>;
        if (isCompressed(frame)) {
          const val = decompressValue(frame.payload);
          payload = (val as Record<string, unknown>) ?? {};
        } else {
          try {
            payload = JSON.parse(
              new TextDecoder().decode(frame.payload),
            ) as Record<string, unknown>;
          } catch {
            continue;
          }
        }

        const msg: JsonRpcMessage = {
          jsonrpc: "2.0",
          ...payload,
        } as JsonRpcMessage;

        this.onmessage?.(msg);
      }
    });
  }

  // ── Send ──────────────────────────────────────────────────────────────

  async send(message: JsonRpcMessage): Promise<void> {
    if (!this.proc?.stdin) throw new Error("Transport not started");

    if (this.useLumen) {
      const frameType =
        message.id !== undefined && message.method
          ? TYPE_REQUEST
          : message.id !== undefined
            ? TYPE_RESPONSE
            : TYPE_NOTIFY;

      const payload = compressValue(
        message as unknown as Record<string, unknown>,
      );
      const total = buildSize(payload.length);
      const buf = new Uint8Array(total);
      buildFrame(frameType, FLAG_COMPRESSED, payload, buf, 0);
      this.proc.stdin.write(Buffer.from(buf));
    } else {
      const line = JSON.stringify(message) + "\n";
      this.proc.stdin.write(line);
    }
  }

  // ── Close ─────────────────────────────────────────────────────────────

  async close(): Promise<void> {
    this.rl?.close();
    this.proc?.stdin?.end();
    this.proc?.kill();
    this.onclose?.();
  }
}


// ═══ WebSocket Transport ════════════════════════════════════════════════════

export interface LumenWebSocketOptions extends LumenTransportOptions {
  /** Send binary WebSocket frames instead of text frames. */
  binaryFrames?: boolean;
}

/**
 * LUMEN over WebSocket (binary frames).
 *
 * Ideal for cloud gateways where the MCP server is behind a WebSocket
 * endpoint (e.g., Cadencia API Gateway → MCP servers).
 *
 * @example
 * ```typescript
 * import { WebSocket } from "ws";
 * import { LumenWebSocketTransport } from "@lumen/mcp-transport";
 *
 * const ws = new WebSocket("wss://api.cadencia.app/mcp");
 * const transport = new LumenWebSocketTransport(ws);
 * await transport.start();
 * ```
 */
export class LumenWebSocketTransport implements Transport {
  private useLumen = false;
  private wsReady = false;
  private assembler = new FrameAssembler();

  onmessage: ((message: JsonRpcMessage) => void) | null = null;
  onerror: ((error: Error) => void) | null = null;
  onclose: (() => void) | null = null;

  constructor(
    private ws: WebSocketLike,
    private options: LumenWebSocketOptions = {},
  ) {}

  // ── Lifecycle ──────────────────────────────────────────────────────────

  async start(): Promise<void> {
    return new Promise((resolve, reject) => {
      // If the WebSocket is already open, negotiate immediately.
      if (
        this.ws.readyState === WebSocketReadyState.OPEN ||
        this.ws.readyState === 1
      ) {
        this.wsReady = true;
        this.negotiateAndSetup(resolve, reject);
        return;
      }

      // Otherwise, wait for the WebSocket to open.
      const originalOnOpen = this.ws.onopen;
      this.ws.onopen = (event: unknown) => {
        if (originalOnOpen) originalOnOpen.call(this.ws, event);
        this.wsReady = true;
        this.negotiateAndSetup(resolve, reject);
      };

      // Bubble errors
      const originalOnError = this.ws.onerror;
      this.ws.onerror = (error: Error) => {
        if (originalOnError) originalOnError.call(this.ws, error);
        this.onerror?.(error);
        reject(new Error("WebSocket connection failed"));
      };

      // Reject if the socket closes before opening
      const originalOnClose = this.ws.onclose;
      this.ws.onclose = (event: { code: number; reason: string }) => {
        if (originalOnClose) originalOnClose.call(this.ws, event);
        if (!this.wsReady) {
          reject(
            new Error(
              `WebSocket closed before negotiation: ${event.code} ${event.reason}`,
            ),
          );
        } else {
          this.onclose?.();
        }
      };
    });
  }

  private negotiateAndSetup(
    resolve: () => void,
    reject: (err: Error) => void,
  ): void {
    if (this.options.forceJsonRpc) {
      this.setupJsonRpcReader();
      resolve();
      return;
    }

    // Send LUMEN probe
    const probe = buildProbe();
    this.ws.send(probe);

    const timeout = this.options.probeTimeoutMs ?? DEFAULT_PROBE_TIMEOUT_MS;
    let negotiated = false;
    const timer = setTimeout(() => {
      if (!negotiated) {
        this.setupJsonRpcReader();
        resolve();
      }
    }, timeout);

    // Use FrameAssembler for robust ACK detection (handles
    // fragmented WebSocket messages where a frame spans multiple
    // onmessage calls).
    const ackAssembler = new FrameAssembler();

    this.ws.onmessage = (event: MessageEventLike) => {
      const data =
        typeof event.data === "string"
          ? new TextEncoder().encode(event.data)
          : event.data instanceof Uint8Array
            ? event.data
            : new Uint8Array(event.data as ArrayBuffer);

      const frames = ackAssembler.push(data);

      for (const frame of frames) {
        if (frame.frameType === 0x10 /* TYPE_PROBE_ACK */) {
          clearTimeout(timer);
          negotiated = true;
          this.useLumen = true;

          const ack = parseAck(frame.payload);
          if (!ack) {
            // ACK frame but malformed payload — treat as failed negotiation
            this.setupJsonRpcReader();
            resolve();
            return;
          }

          // Negotiation succeeded — switch to LUMEN reader.
          // Any leftover bytes in ackAssembler are discarded;
          // the server shouldn't send anything before the ACK.
          this.setupLumenReader();
          resolve();
        }
      }
    };
  }

  private setupJsonRpcReader(): void {
    this.ws.onmessage = (event: MessageEventLike) => {
      const text =
        typeof event.data === "string"
          ? event.data
          : new TextDecoder().decode(event.data as ArrayBuffer);
      try {
        const msg = JSON.parse(text) as JsonRpcMessage;
        this.onmessage?.(msg);
      } catch {
        // Ignore
      }
    };
  }

  private setupLumenReader(): void {
    this.ws.onmessage = (event: MessageEventLike) => {
      const raw =
        event.data instanceof Uint8Array
          ? new Uint8Array(
              event.data.buffer,
              event.data.byteOffset,
              event.data.byteLength,
            )
          : typeof event.data === "string"
            ? new TextEncoder().encode(event.data)
            : new Uint8Array(event.data as ArrayBuffer);

      const frames = this.assembler.push(raw);

      for (const frame of frames) {
        let payload: Record<string, unknown>;
        if (isCompressed(frame)) {
          const val = decompressValue(frame.payload);
          payload = (val as Record<string, unknown>) ?? {};
        } else {
          try {
            payload = JSON.parse(
              new TextDecoder().decode(frame.payload),
            ) as Record<string, unknown>;
          } catch {
            continue;
          }
        }

        const msg: JsonRpcMessage = {
          jsonrpc: "2.0",
          ...payload,
        } as JsonRpcMessage;

        this.onmessage?.(msg);
      }
    };
  }

  // ── Send ──────────────────────────────────────────────────────────────

  async send(message: JsonRpcMessage): Promise<void> {
    if (!this.wsReady) throw new Error("Transport not started");

    if (this.useLumen) {
      const frameType =
        message.id !== undefined && message.method
          ? TYPE_REQUEST
          : message.id !== undefined
            ? TYPE_RESPONSE
            : TYPE_NOTIFY;

      const payload = compressValue(
        message as unknown as Record<string, unknown>,
      );
      const total = buildSize(payload.length);
      const buf = new Uint8Array(total);
      buildFrame(frameType, FLAG_COMPRESSED, payload, buf, 0);
      this.ws.send(buf);
    } else {
      this.ws.send(JSON.stringify(message));
    }
  }

  // ── Close ─────────────────────────────────────────────────────────────

  async close(): Promise<void> {
    // Close the WebSocket if it's still open/connecting.
    if (
      this.ws.readyState === WebSocketReadyState.OPEN ||
      this.ws.readyState === WebSocketReadyState.CONNECTING ||
      this.ws.readyState === 1 ||
      this.ws.readyState === 0
    ) {
      this.ws.close(1000, "Transport closed");
    }
    this.wsReady = false;
    this.onclose?.();
  }
}

// ═══ WebSocket-like interface ═══════════════════════════════════════════════

export enum WebSocketReadyState {
  CONNECTING = 0,
  OPEN = 1,
  CLOSING = 2,
  CLOSED = 3,
}

export interface MessageEventLike {
  data: string | ArrayBuffer | Uint8Array;
}

export interface WebSocketLike {
  readyState: WebSocketReadyState | number;
  send(data: string | Uint8Array | ArrayBuffer): void;
  close(code?: number, reason?: string): void;
  onopen: ((event: unknown) => void) | null;
  onmessage: ((event: MessageEventLike) => void) | null;
  onerror: ((error: Error) => void) | null;
  onclose: ((event: { code: number; reason: string }) => void) | null;
}

File: implementations/python/src/lumen/transport.py
"""
Transport implementations — drop-in replacements for MCP SDK transports.

Each transport performs automatic LUMEN negotiation on ``start()``.
If the remote peer doesn't respond to the LUMEN probe within a
configurable timeout, the transport falls back to JSON-RPC transparently.

Ported from TypeScript ``src/transport.ts``.
"""

from __future__ import annotations

import asyncio
import json
import logging
import subprocess
import os
import sys
import threading
from abc import ABC, abstractmethod
from typing import Any, Callable

from .compress import compress_value, decompress_value
from .frame import (
    FLAG_COMPRESSED,
    TYPE_NOTIFY,
    TYPE_PROBE_ACK,
    TYPE_REQUEST,
    TYPE_RESPONSE,
    build_frame,
    is_compressed,
)
from .frame_assembler import FrameAssembler
from .negotiation import (
    DEFAULT_PROBE_TIMEOUT_MS,
    build_probe,
    parse_ack,
)

logger = logging.getLogger(__name__)



def _flush_stdin(process: "subprocess.Popen[bytes]") -> None:
    """Cross-platform stdin flush. Windows raises OSError on pipe flush."""
    try:
        process.stdin.flush()
    except OSError:
        pass  # Windows: flushing stdin pipe may fail

# ═══ Message types ═══════════════════════════════════════════════════════════

# Matches the MCP SDK's JSONRPCMessage shape
JsonRpcMessage = dict[str, Any]
OnMessageCallback = Callable[[JsonRpcMessage], None] | None
OnErrorCallback = Callable[[Exception], None] | None
OnCloseCallback = Callable[[], None] | None


# ═══ Transport interface ═════════════════════════════════════════════════════


class Transport(ABC):
    """Minimal transport interface compatible with MCP SDK's ``Transport``."""

    @abstractmethod
    async def start(self) -> None:
        """Start the transport and perform LUMEN negotiation."""

    @abstractmethod
    async def send(self, message: JsonRpcMessage) -> None:
        """Send a JSON-RPC message to the remote peer."""

    @abstractmethod
    async def close(self) -> None:
        """Close the transport gracefully."""

    onmessage: OnMessageCallback = None
    onerror: OnErrorCallback = None
    onclose: OnCloseCallback = None


# ═══ Stdio Transport ═════════════════════════════════════════════════════════


class LumenStdioTransport(Transport):
    """Drop-in replacement for MCP SDK's ``StdioServerParameters``-based transport.

    On ``start()``, spawns the MCP server process and attempts LUMEN negotiation.
    If the server responds with a LUMEN ACK frame, all subsequent communication
    uses LUMEN binary frames. Otherwise, falls back to JSON-RPC line protocol.

    Usage::

        transport = LumenStdioTransport(
            command="python", args=["my_mcp_server.py"]
        )
        await transport.start()
        # Now use transport.send() / transport.onmessage like any MCP transport
    """

    def __init__(
        self,
        command: str,
        args: list[str] | None = None,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
        *,
        force_json_rpc: bool = False,
        probe_timeout_ms: float = DEFAULT_PROBE_TIMEOUT_MS,
    ) -> None:
        self._command = command
        self._args = args or []
        self._env = env
        self._cwd = cwd
        self._force_json_rpc = force_json_rpc
        self._probe_timeout_ms = probe_timeout_ms / 1000.0

        self._process: subprocess.Popen[bytes] | None = None
        self._use_lumen = False
        self._assembler = FrameAssembler()
        self._reader_task: asyncio.Task[None] | None = None
        self._stderr_task: asyncio.Task[None] | None = None
        self._closed = False
        # Chunk reader: a dedicated thread pushes stdout chunks into an
        # asyncio queue to avoid O(threadpool) byte-at-a-time dispatch.
        self._chunk_queue: asyncio.Queue[bytes] = asyncio.Queue(maxsize=256)
        self._stdout_thread: threading.Thread | None = None

        # Callbacks
        self.onmessage: OnMessageCallback = None
        self.onerror: OnErrorCallback = None
        self.onclose: OnCloseCallback = None

    # ── Lifecycle ──────────────────────────────────────────────────────────

    async def start(self) -> None:
        """Spawn the server process and negotiate LUMEN."""
        # On Windows, asyncio subprocess needs different setup
        kwargs: dict[str, Any] = {
            "args": [self._command, *self._args],
            "stdin": subprocess.PIPE,
            "stdout": subprocess.PIPE,
            "stderr": subprocess.PIPE,
            "cwd": self._cwd,
        }
        if self._env:
            kwargs["env"] = {**os.environ, **self._env}

        # Use regular subprocess for simplicity; Python's asyncio subprocess
        # has quirks on Windows. We'll use a thread-based reader instead.
        self._process = subprocess.Popen(**kwargs)

        # Start the stdout reader thread (feeds chunks into async queue)
        self._start_stdout_thread()

        # Start stderr passthrough
        if self._process.stderr:
            self._stderr_task = asyncio.create_task(self._log_stderr())

        # ── LUMEN negotiation ──────────────────────────────────────────
        if not self._force_json_rpc:
            probe = build_probe()
            assert self._process.stdin
            self._process.stdin.write(probe.to_bytes())
            _flush_stdin(self._process)

            if await self._wait_for_ack():
                self._use_lumen = True
                self._reader_task = asyncio.create_task(self._read_lumen())
                return

        # ── JSON-RPC fallback ──────────────────────────────────────────
        self._reader_task = asyncio.create_task(self._read_jsonrpc())

    async def send(self, message: JsonRpcMessage) -> None:
        """Send a message to the remote peer."""
        if not self._process or not self._process.stdin:
            raise RuntimeError("Transport not started")

        if self._use_lumen:
            self._send_lumen(message)
        else:
            self._send_jsonrpc(message)

    async def close(self) -> None:
        """Close the transport and terminate the child process."""
        if self._closed:
            return
        self._closed = True

        # Close stdin FIRST so the child process exits and the reader
        # tasks (blocked on stdout/stderr.read()) can complete.
        if self._process and self._process.stdin:
            try:
                self._process.stdin.close()
            except Exception:
                pass

        for task in (self._stderr_task, self._reader_task):
            if task:
                task.cancel()
                try:
                    await task
                except asyncio.CancelledError:
                    pass
        self._stderr_task = None
        self._reader_task = None

        if self._process:
            self._process.terminate()
            try:
                self._process.wait(timeout=3)
            except subprocess.TimeoutExpired:
                self._process.kill()
                self._process.wait()
            self._process = None

        if self.onclose:
            self.onclose()

    # ── Send helpers ───────────────────────────────────────────────────────

    def _send_jsonrpc(self, message: JsonRpcMessage) -> None:
        assert self._process and self._process.stdin
        line = json.dumps(message, ensure_ascii=False) + "\n"
        self._process.stdin.write(line.encode("utf-8"))
        _flush_stdin(self._process)

    def _send_lumen(self, message: JsonRpcMessage) -> None:
        assert self._process and self._process.stdin
        # Determine frame type from message shape
        if "method" in message and "id" in message:
            frame_type = TYPE_REQUEST
        elif "result" in message or "error" in message:
            frame_type = TYPE_RESPONSE
        else:
            frame_type = TYPE_NOTIFY

        payload = compress_value(message)
        frame = build_frame(frame_type, FLAG_COMPRESSED, payload)
        self._process.stdin.write(frame.to_bytes())
        _flush_stdin(self._process)

    # ── Receive helpers ────────────────────────────────────────────────────

    def _start_stdout_thread(self) -> None:
        """Launch a dedicated thread that reads stdout in 64 KiB chunks.

        Chunks are pushed into ``_chunk_queue`` via the event loop so the
        async reader can consume them without per-byte threadpool dispatch.
        On EOF the thread pushes ``b""`` as a sentinel.
        """
        loop = asyncio.get_running_loop()
        stdout = self._process.stdout  # captured before thread starts

        def _run() -> None:
            try:
                while True:
                    # 64 KiB chunks — large enough to amortize syscall cost,
                    # small enough to keep latency acceptable.
                    chunk = stdout.read(65536)
                    if chunk is not None:
                        loop.call_soon_threadsafe(self._chunk_queue.put_nowait, chunk)
                    if not chunk:
                        break  # EOF
            except Exception:
                # Pipe closed or other I/O error — push sentinel so the
                # async reader can clean up.
                try:
                    loop.call_soon_threadsafe(self._chunk_queue.put_nowait, b"")
                except Exception:
                    pass

        self._stdout_thread = threading.Thread(target=_run, daemon=True)
        self._stdout_thread.start()

    async def _read_lumen(self) -> None:
        """Read binary LUMEN frames from the chunk queue (fed by reader thread)."""
        while True:
            try:
                chunk = await self._chunk_queue.get()
            except Exception as exc:
                if self.onerror:
                    self.onerror(exc)
                return

            if not chunk:
                # EOF sentinel
                if not self._closed and self.onclose:
                    self.onclose()
                return

            frames = self._assembler.push(chunk)
            for frame in frames:
                try:
                    payload = decompress_value(frame.payload) if is_compressed(frame) else json.loads(frame.payload)
                    if isinstance(payload, dict) and self.onmessage:
                        self.onmessage(payload)
                except Exception as exc:
                    if self.onerror:
                        self.onerror(exc)

    async def _read_jsonrpc(self) -> None:
        """Read JSON-RPC line-delimited messages from the chunk queue."""
        # Buffer for partial lines
        buf = b""

        while True:
            try:
                chunk = await self._chunk_queue.get()
            except Exception as exc:
                if self.onerror:
                    self.onerror(exc)
                return

            if not chunk:
                # Partial line at EOF is normal
                if buf.strip():
                    try:
                        message = json.loads(buf)
                        if self.onmessage:
                            self.onmessage(message)
                    except json.JSONDecodeError:
                        pass
                if not self._closed and self.onclose:
                    self.onclose()
                return

            buf += chunk
            while b"\n" in buf:
                line, buf = buf.split(b"\n", 1)
                line = line.strip()
                if not line:
                    continue
                try:
                    message = json.loads(line)
                    if self.onmessage:
                        self.onmessage(message)
                except json.JSONDecodeError:
                    pass

    async def _log_stderr(self) -> None:
        """Passthrough stderr for debugging."""
        assert self._process and self._process.stderr
        loop = asyncio.get_running_loop()

        while True:
            try:
                chunk = await loop.run_in_executor(
                    None, self._process.stderr.readline
                )
            except Exception:
                return
            if not chunk:
                return
            logger.debug("[mcp-server stderr] %s", chunk.decode("utf-8", errors="replace"))

    async def _wait_for_ack(self) -> bool:
        """Wait for a PROBE_ACK frame via the chunk queue (fed by reader thread)."""
        loop = asyncio.get_running_loop()
        deadline = loop.time() + self._probe_timeout_ms

        while loop.time() < deadline:
            remaining = deadline - loop.time()
            if remaining <= 0:
                break

            try:
                chunk = await asyncio.wait_for(
                    self._chunk_queue.get(),
                    timeout=min(remaining, 0.1),
                )
            except asyncio.TimeoutError:
                continue
            except Exception:
                return False

            if not chunk:
                return False  # EOF

            frames = self._assembler.push(chunk)
            for frame in frames:
                if frame.frame_type == TYPE_PROBE_ACK:
                    try:
                        value = decompress_value(frame.payload)
                        return isinstance(value, dict) and "server_name" in value
                    except Exception:
                        return False

        return False


# ═══ WebSocket Transport ═════════════════════════════════════════════════════


class LumenWebSocketTransport(Transport):
    """WebSocket transport with LUMEN binary frames.

    Ideal for cloud gateways (Cadencia → API Gateway → MCP servers).

    Usage::

        transport = LumenWebSocketTransport("ws://localhost:8080/mcp")
        await transport.start()
    """

    def __init__(
        self,
        url: str,
        *,
        force_json_rpc: bool = False,
        probe_timeout_ms: float = DEFAULT_PROBE_TIMEOUT_MS,
        binary_frames: bool = True,
    ) -> None:
        self._url = url
        self._force_json_rpc = force_json_rpc
        self._probe_timeout_ms = probe_timeout_ms / 1000.0
        self._binary_frames = binary_frames

        self._ws: Any = None  # websockets connection
        self._use_lumen = False
        self._assembler = FrameAssembler()
        self._receiver_task: asyncio.Task[None] | None = None

        self.onmessage: OnMessageCallback = None
        self.onerror: OnErrorCallback = None
        self.onclose: OnCloseCallback = None

    async def start(self) -> None:
        """Connect WebSocket and negotiate LUMEN."""
        try:
            import websockets
        except ImportError:
            raise ImportError(
                "websockets package required for LumenWebSocketTransport. "
                "Install with: pip install websockets"
            )

        self._ws = await websockets.connect(self._url)

        # ── LUMEN negotiation ──────────────────────────────────────────
        if not self._force_json_rpc:
            probe = build_probe()
            await self._ws.send(probe.to_bytes())
            try:
                ack_data = await asyncio.wait_for(
                    self._ws.recv(), timeout=self._probe_timeout_ms
                )
            except asyncio.TimeoutError:
                pass  # fallback to JSON-RPC
            else:
                if isinstance(ack_data, bytes):
                    ack = parse_ack(ack_data)
                    if ack:
                        self._use_lumen = True

        self._receiver_task = asyncio.create_task(self._receive_ws())

    async def send(self, message: JsonRpcMessage) -> None:
        if not self._ws:
            raise RuntimeError("Transport not started")

        if self._use_lumen:
            payload = compress_value(message)
            frame = build_frame(TYPE_REQUEST, FLAG_COMPRESSED, payload)
            await self._ws.send(frame.to_bytes())
        else:
            await self._ws.send(json.dumps(message, ensure_ascii=False))

    async def close(self) -> None:
        if self._receiver_task:
            self._receiver_task.cancel()
            try:
                await self._receiver_task
            except asyncio.CancelledError:
                pass

        if self._ws:
            await self._ws.close()
            self._ws = None

        if self.onclose:
            self.onclose()

    async def _receive_ws(self) -> None:
        assert self._ws
        try:
            async for message in self._ws:
                if isinstance(message, bytes):
                    frames = self._assembler.push(message)
                    for frame in frames:
                        try:
                            payload = (
                                decompress_value(frame.payload)
                                if is_compressed(frame)
                                else json.loads(frame.payload)
                            )
                            if isinstance(payload, dict) and self.onmessage:
                                self.onmessage(payload)
                        except Exception as exc:
                            if self.onerror:
                                self.onerror(exc)
                else:
                    try:
                        parsed = json.loads(message)
                        if self.onmessage:
                            self.onmessage(parsed)
                    except json.JSONDecodeError:
                        pass
        except Exception as exc:
            if self.onerror:
                self.onerror(exc)
        finally:
            if self.onclose:
                self.onclose()

# ── Backward compatibility ──────────────────────────────────
# Hermes Agent PR #47740 references MCPLumenTransport.
# Keep this alias until Hermes updates its import.
MCPLumenTransport = LumenStdioTransport
